### ISSUE TITLE
BOLUS CALCULATOR on Xdrip

### ISSUE THREAD

**Issue by Marcio261** [NightscoutFoundation/xDrip#1479]
Hello!
Does anyone know if there is a function of bolus calculator on the app?? If it isn't available, i think it is a way to make a more complete app, using the "extra definitions/simulation ratio...." To Database, and on the "add painel"(inicial screen) shows(next to the other icons) an icon option for calculador, inputing the total carbs of the meal, and it shows the dose (using relation ratio carbs - insulin) .But its also a suggestion.
Thank you.

**Comment by doc77** [comment]
It's already there. Just enter your carbs (tap the + button to make sure they are entered). Then, based on # of carbs, current Bg, direction and force of momentum, etc., you will get your bolus estimation displayed on the top of the main screen.

**Comment by Marcio261** [comment]
Ow, you mean the simulation right? That shows the preview graphic of the glucose(based on insulin-carbs)?
Ok, in my case, i take different doses of insulin during the day(depending the hour), and i entered those different doses on the simulation settings, and every time i ate, i need always to go to the settings to see how much i should take for each portion, then go to the calculator(other app) to make my maths hahaha. But if the app had already a calculator on it(that already know how much insulin i take on that period of the day), that i simply input the # of carbs and the app suggests how much insulin i should take.
Anyway, sorry for this big text hahaha.
And thank you.

**Comment by tolot27** [comment]
@Marcio261 Why did you remove your comment and closed this issue?

I believe you missed something @doc77 is referring to. If you enter the amount of carbs and save the value, xDrip will show the amount of insulin needed below on top of the page below the xDrip+ letters. I'll post a screenshot later, if I have access to a device.

**Comment by Marcio261** [comment]
Yes yes I saw the comment of @doc77 . which is why I removed my other comment and closed the question because I thought it would no longer be necessary to leave it open. Sorry hahaha. But I already found the setting, and I couldn't see it beacause I had to enable the "show insulin / carbo ratio" option. Thanks a lot for the help! :)

**Comment by ShusterRooster** [comment]
@tolot27 @Marcio261 oh wow, I started working on a bolus calculator for xDrip+ a few days ago and I wanted to see if there was a feature request for it already. I've gotten a decent amount of it done and I'm planning on adding even more things. Once my code becomes more stable and optimized and whatnot, I plan on making a pull request. 

Here's what I've gotten done so far:
https://imgur.com/a/jRdHGed

I would love to hear your thoughts on what other features I should add to this!

**Comment by Marcio261** [comment]
@ShusterRooster I really liked what I saw in the video. I especially liked the separation between the bolus of meals and the corrections ... I really liked the detail! 😉 I only have one question: is it always necessary to enter the blood glucose value? (I understand if it is in the case of hyperglycemia and having to measure capillary blood glucose. But these are details, I think) And a suggestion that I personally like are "widgets" haha ​​... I find it very practical. It would be a simple widget. Something like a calculator widget 🤔 just insert the carbs and the suggestion is shown.❤️ And congratulations also for the initiative 😊 we diabetics thank you hahaha😉

**Comment by ShusterRooster** [comment]
@Marcio261 Yeah I can definitely have an option where you don't need to enter a bg value. As of right now, It just pulls whatever your current bg is and places that in the text field initially. I will have a settings option to enable a widget as well now that you've mentioned it. Thanks for the feedback and support!

**Comment by Marcio261** [comment]
@ShusterRooster Exactly. I think they would be good app upgrades💪. If you need any more feedback from the user's perspective, you can count on me.

**Comment by TheDiabeticWay** [comment]
I was thinking of a different feature but still related to bolusing. Maybe we can somehow have a scan icon to scan barcodes like myfitnesspal does but from withing xdrip where we add carbs. I m already adding recipes to a site with bar codes on so would be great for this as well as for it going to myfitnesspal and giving a return carbs for us to accept this will make it a lot easier for a quicker calculating carbs up from myfitnesspal or other software too. Any thoughts if this can be possible

**Comment by ShusterRooster** [comment]
@TheDiabeticWay I'm not sure I'd be able to do that, I'm nowhere near experienced enough to implement that yet. I can ask around and see if there's enough interest to gather some people up and help me make the db integration for whatever service we're hooking up to.

I submitted my code just a bit ago so we'll have to see what the devs say in terms of actually getting this thing into the repo.
Here's my PR: https://github.com/NightscoutFoundation/xDrip/pull/1589

**Comment by Navid200** [comment]
@Marcio261 Would you please edit the title and remove all caps?
